### DISCUSSION TITLE
type of an assigned __doc__ is `str|None`

### DISCUSSION THREAD

**Discussion by bukzor** [microsoft/pyright#3820] [Q&A]
This is my example program:

`demo.py`:
```python
"""Show a given range of characters."""
def main():
    reveal_type(__doc__)  # T: Type of "__doc__" is "str | None"
    return "Usage: " + __doc__  # E: Operator "+" not supported for types "Literal['Usage: ']" and "str | None"  Operator "+" not supported for types "Literal['Usage: ']" and "None" Pyright (reportGeneralTypeIssues) [6, 12]
if __name__ == "__main__":
    exit(main())
```

Running the program shows that the type is certainly `str` and can't be `None`.

```
$ python demo.py 
Usage: Show a given range of characters.
```

Is the `| None` inference here a feature or a bug?

**Comment by erictraut** [comment]
The type of the module-level symbol `__doc__` is not inferred. It is hard-coded in pyright's logic along with several other module-level attributes such as `__name__`, `__qualname__`, `__path__`, etc. These standard module attributes unfortunately aren't defined anywhere in typehsed stubs, so type checkers need to hard-code them. In contrast, `__doc__` is defined in the typeshed stubs for the `object` class. Its type is declared as `str | None` there.

So I don't consider this a bug. It's "as designed".

**Comment by bukzor** [comment]
> So I don't consider this a bug. ...

Can we agree that people familiar with python but not the implementation of python typechecking would characterize a "cannot concatenate None and str" error on this line as strictly incorrect?

I think this is closer to the heart of the issue: I understand these two programs (below) to be exactly equivalent, and indeed when run they have exactly the same behavior, but pyright sees them as very different.

```python
__doc__ = """x"""
print(__doc__)
reveal_type(__doc__)  # Literal['x']
```

```python
"""x"""
print(__doc__)
reveal_type(__doc__)  # None | str
```

While pyright sees this as no different:

```python
print(__doc__)
reveal_type(__doc__)  # None | str
```

> It's "as designed".

I agree there's no implementation bug here, but there are design/UX bugs too. To eliminate those bug categories we also have to consider whether the design accomplishes the intent, and whether an acceptable percentage of the intended audience would find it productive. I don't know enough of your intent to speak to the first, and the second is largely a matter of prioritization, so I have to ask you for it. Once I know, I can bother you with irrelevant discussion less :)

I've found pyright to be an impressive and extremely helpful tool, and so these little things stick out to me. I'd be delighted to improve it in even a small way, but if you believe that there is no change here that could be an improvement I'll move along.

**Reply by erictraut** [reply]
I agree that there is room for improvement in usability for this particular case through the addition of special-case logic that dynamically adjusts the type of the module's `__doc__` type. We generally try to avoid such special-case logic because it leads to complexity and maintainability issues. We do make exceptions in common cases though. Since this is the first time anyone has reported this issue, I'm not convinced it's a common case. It may be worth considering as a future enhancement if this discussion thread receives more upvotes.

**Reply by bukzor** [reply]
Thank you for bearing with me :) I'm aware that I'm not the most pleasant to deal with, and I appreciate your time.

> I agree that there is room for improvement in usability for this particular case through the addition of special-case logic that dynamically adjusts the type of the module's __doc__ type.

I don't understand why it must be a special case. I understand this as a general case of assigning a value to a variable. 

I suppose the special case is that the assignment of `__doc__` looks very odd with respect to most assignments? But assignment of class, function, and method variables are also odd in a similar(ish) fashion.

Should I instead have a look at the code? Perhaps that will give me a more similar point of view to yours. I'll attempt to improve this without any sacrifice to generality. My goal would be to teach pyright that docstring declarations are no different from assignments to a `__doc__` variable, which would eliminate any and all need for special-case code.

**Reply by erictraut** [reply]
Normal variable assignments occur at an identifiable point within a code flow graph. The implicit assignment of `__doc__` by the module import loader is done outside of any code flow graph. They're very different.

As I said, this would require special-case logic to handle, and I'm not interested in adding such special-case logic based on one user reporting this issue. If we receive sufficient signal from other users that this is important and generally useful, we may reconsider.

**Reply by bukzor** [reply]
> Normal variable assignments occur at an identifiable point within a code flow graph. The implicit assignment of `__doc__` by the module import loader is done outside of any code flow graph ...

That may be true as a matter of implementation, but as a matter of semantics (with minimum special cases) a docstring statement is exactly equivalent to a `__doc__` assignment. This definition admits docstrings for functions and classes with no special cases (since there is certainly no module import loader involved there).

...

But also as a matter of implementation a docstring is exactly a `__doc__` assignment:

```sh
$ cat demo.py 
"""3"""

$ python -c 'import dis; import marshal; import demo; dis.dis(marshal.loads(open(demo.__cached__, "rb").read()[16:]))'
  1           0 LOAD_CONST               0 ('3')
              2 STORE_NAME               0 (__doc__)
              4 LOAD_CONST               1 (None)
              6 RETURN_VALUE
              
$ echo '__doc__ = """4"""' | tee demo.py 
__doc__ = """4"""

$ python -c 'import dis; import marshal; import demo; dis.dis(marshal.loads(open(demo.__cached__, "rb").read()[16:]))'
  1           0 LOAD_CONST               0 ('4')
              2 STORE_NAME               0 (__doc__)
              4 LOAD_CONST               1 (None)
              6 RETURN_VALUE

```

It even has line number 1 attached.

**Reply by erictraut** [reply]
And that point would be relevant if static type analyzers worked from byte codes, but they don't.

**Reply by bukzor** [reply]
I want to write a small patch that corrects the semantics of `__doc__` in pyright by interpreting docstring statements as an assignment to `__doc__` at the line of the docstring statement. 

This is a (tiny) improvement to accuracy because it better models the everyday semantics of python (as well as the cpython implementation), so I hope you would consider accepting such a change.

Will you consider that change? Given that it must have no more than a handful of "special case" lines of code, and it will come with tests to bring the maintenance burden to near-zero. I appreciate that few people will benefit from such a small improvement, and that's why I'm volunteering. My current goal is only to get agreement from approvers that this *could* be a good and proper change before I start work on it. 


P.S.

> And that point would be relevant if static type analyzers worked from byte codes, but they don't.

This isn't relevant, but you may be interested to learn that pytype does actually use the bytecode: https://google.github.io/pytype/developers/index.html#basic-concepts

**Reply by erictraut** [reply]
We don't incorporate new functionality in pyright just because one user thinks it's a good idea. If this feature is important enough to you, you're of course welcome to fork pyright and modify it as you would like. And we can see if we get signal from other pyright users (in the form of additional thumbs-up votes in this discussion thread) to see if this functionality is something that is more generally desired.

**Reply by erictraut** [reply]
If we eventually add support for this, it will involve some unnatural hacks and code that is not well aligned with the overall pyright architecture. It will also create an inconsistency between the behavior of the module `__doc__` attribute and the attribute of the same name within classes and functions. This which will invite questions about the latter, which will potentially lead to more unnatural hacks.

There are cases where a proposed feature is simple and straightforward to add, but this isn't one of those.

**Reply by bukzor** [reply]
> ... It will also create an inconsistency between the behavior of the module __doc__ attribute and the attribute of the same name within classes and functions.

They'll all three use the same code and be fixed simultaneously. I tried to explain why above. I'll just show it. I understand you have no intention to merge it but I believe you'll find it's not the monstrosity you think it is.
